Sming/Components/crypto/include/Crypto/HashContext.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <axtls-8266/crypto/crypto.h>
+#include <WString.h>
+#include <Data/HexString.h>
+
+namespace Crypto
+{
+/**
+ * @brief
+ */
+struct Blob {
+	const void* data;
+	size_t length;
+
+	Blob(const void* data, size_t length) : data(data), length(length)
+	{
+	}
+
+	Blob(const String& str) : data(str.c_str()), length(str.length())
+	{
+	}
+};
+
+/**
+ * @brief Class template for a Hash value, essentially just an array of bytes
+ */
+template <size_t size> struct HashValue {
+	uint8_t data[size];
+
+	static constexpr size_t length = size;
+
+	bool copy(const void* src)
+	{
+		if(src == nullptr) {
+			memset(data, 0, size);
+			return false;
+		}
+
+		memcpy(data, src, size);
+		return true;
+	}
+
+	bool operator==(const HashValue& other) const
+	{
+		return memcmp(data, other.data, sizeof(data)) == 0;
+	}
+
+	String toString(char separator = '\0') const
+	{
+		return makeHexString(data, size, separator);
+	}
+};
+
+/**
+ * @brief Class template for a Hash implementation 'Context'
+ */
+template <class Context, class Hash> class HashContext
+{
+public:
+	/**
+	 * @brief Calculate hash on a single block of data
+	 * @retval Hash
+	 * @{
+	 */
+	static Hash calculate(const void* data, size_t length)
+	{
+		Context ctx;
+		ctx.update(data, length);
+		return ctx.hash();
+	}
+
+	template <typename T> static Hash calculate(const T& blob)
+	{
+		Context ctx;
+		ctx.update(blob);
+		return ctx.hash();
+	}
+	/** @} */
+
+	/**
+	 * @brief Update hash over a given block of data
+	 * @{
+	 */
+	void update(const Blob& blob)
+	{
+		static_cast<Context*>(this)->update(blob.data, blob.length);
+	}
+
+	void update(const String& str)
+	{
+		static_cast<Context*>(this)->update(str.c_str(), str.length());
+	}
+
+	/*
+	 * Update from a fixed array or struct, but not a pointer - base method handles that, with size
+	 */
+	template <typename T> typename std::enable_if<!std::is_pointer<T>::value, void>::type update(const T& data)
```

Comment:
Use `enable_if_t` and `is_pointer_v` instead.
Maybe also constrain on `is_pod`?

Reply:
For now, keeping to C++11 so have to stick with existing traits.
Re-checking `is_pod` appears to be deprecated in C++20. I _think_ we need `is_std_layout`?
https://stackoverflow.com/questions/6496545/trivial-vs-standard-layout-vs-pod

Reply:
(I know, I've used it in FlashString - something to fix.)

Reply:
Handn't noticed that yet. `is_std_layout` seems right then.

Reply:
I've removed this method, it's just asking for trouble.